File: dart/src/queue/Queue.dart
class Node {
  int? data;
  Node? next;

  Node(this.data);
}

class Queue {
  Node? front;
  Node? rear;
  int length = 0;
  Queue() {
    this.front = null;
    this.rear = null;
    this.length = 0;
  }
  void enqueue(int data) {
    var newNode = Node(data);
    if (this.front == null) {
      this.front = newNode;
    }
    if (this.rear != null) {
      this.rear!.next = newNode;
    }
    this.rear = newNode;
    this.length = this.length + 1;
  }

  int? dequeue() {
    if (this.front == null) {
      return null;
    }
    var node = this.front;
    this.front = this.front?.next;
    if (this.front == null) {
      this.rear = null;
    }
    this.length = this.length - 1;
    return node?.data;
  }

  int? peek() {
    if (this.front == null) {
      return null;
    }
    return this.front?.data;
  }

  bool isEmpty() {
    return this.front == null;
  }
}

File: go/queue/queue.go
package queue

type Node struct {
	data int
	next *Node
}

type Queue struct {
	front  *Node
	rear   *Node
	length int
}

func New() *Queue {
	return &Queue{
		front:  nil,
		rear:   nil,
		length: 0,
	}
}

func (q *Queue) Enqueue(data int) {
	node := &Node{
		data: data,
		next: nil,
	}
	if q.front == nil {
		q.front = node
	} else {
		q.rear.next = node
	}
	q.rear = node
	q.length++
}

func (q *Queue) Dequeue() int {
	if q.front == nil {
		return -1
	}
	data := q.front.data
	q.front = q.front.next
	if q.front == nil {
		q.rear = nil
	}
	q.length--
	return data
}

func (q *Queue) Size() int {
	return q.length
}
func (q *Queue) IsEmpty() bool {
	return q.front == nil
}
func (q *Queue) Print() {
	if q.front == nil {
		return
	}
	for node := q.front; node != nil; node = node.next {
		print(node.data, " ")
	}
	println()
}
